pkg/kubelet/images/image_manager.go

```diff
@@ -57,21 +76,35 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 		imageService: imageService,
 		backOff:      imageBackOff,
 		puller:       puller,
+		keyring:      keyring,
 	}
 }
 
 // shouldPullImage returns whether we should pull an image according to
 // the presence and pull policy of the image.
-func shouldPullImage(container *v1.Container, imagePresent bool) bool {
+func shouldPullImage(container *v1.Container, imagePresent, pulledBySecret, ensuredBySecret bool) bool {
 	if container.ImagePullPolicy == v1.PullNever {
 		return false
```

Comment:
consider the case 
```
PodA
  image: imageA
  imagePullPolicy: pullIfNotPresent
  pullImageSecret: ###
```

```
PodB
 image: imageA
 imagePullPolicy: Never
```
with image pull policy `never`, Pod B will be able to access  imageA, shouldn't we handle this case also?

Reply:
definitely should handle that case to.. can't pull though.. so result is still false here and below on line ..

ensuredBySecret will be false for this case and if pulledBySecret we have a new error case...

so below in this snippet on line 161:
```
		if present {
 			msg := fmt.Sprintf("Container image %q already present on machine", container.Image)
 			m.logIt(ref, v1.EventTypeNormal, events.PulledImage, logPrefix, msg, klog.Info)
			return imageRef, "", nil
		}
``` 
we'll have to add an error:
```
if present {
    if pulledBySecret && !ensuredBySecret {
	msg := fmt.Sprintf("Container image %q is present with pull policy of Never but does not have the proper permission (image secret) that was used to pull the image", container.Image)
	m.logIt(ref, v1.EventTypeWarning, events.ErrImageNeverPullPolicy, logPrefix, msg, klog.Warning)
	return "", msg, ErrImageNeverPull
   }
   msg := fmt.Sprintf("Container image %q already present on machine", container.Image)
   m.logIt(ref, v1.EventTypeNormal, events.PulledImage, logPrefix, msg, klog.Info)
   return imageRef, "", nil
}

```